### ISSUE TITLE
Provider is sent kube-system pods

### ISSUE THREAD

**Issue by peterhuene** [deislabs/krustlet#167]
# Steps to reproduce

1. Run `krustlet-wasi`:

```
RUST_LOG="info,kubelet=trace,cranelift_wasm=warn" cargo run --release --bin krustlet-wasi
```

# Expected results

The WASI provider should only be given pods matching the WASI architecture.

# Actual results

The provider is given pods from `kube-system`, which fails to fetch the modules of, resulting in this error:

```text
[2020-04-07T23:03:24Z ERROR kubelet::kubelet] Error handling event: error sending request for url (https://docker/v2/): error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known
```

In this instance, the pod was `kube-system/storage-provisioner` which has a container image of `docker/desktop-storage-provisioner:v1.0`.  Obviously that's not an OCI image that would work for the provider anyway.

This appears related to the changes in #151 which altered how pods get scheduled onto the krustlet node.

**Comment by peterhuene** [comment]
I do see the new taint on the node, so I'm not sure why these pods are being scheduled:

```
Taints:             krustlet/arch=wasm32-wasi:NoExecute
```

**Comment by thomastaylor312** [comment]
@peterhuene Yeah, this is a known issue and I am not quite sure how to prevent it. This is coming because those daemonset pods (like kube-proxy) have tolerations set that allow anything. I am trying to avoid making user set both a node selector and a toleration. We need the taint on the node to avoid scheduling other pods, so we'd have to have both. 

Right now we are just letting those pods error as indicated in the release notes:
> Unsupported workloads (such as those dropped automatically onto a node like kube-proxy) can get into an error loop. This is more of a nuisance that will cause some logging noise, but not impact the running of Krustlet

So if anyone out there has some ideas of what we can do to handle this, we are all ears

**Comment by DazWilkin** [comment]
I've spent some time this week deploying Krustlet to GKE, DigitalOcean (DO) and Linode and each of these manifests some form of this issue. I forget the list with GKE as I'm struggling to get Bootstrap Tokens working, but DO has this [list](https://pretired.dazwilkin.com/posts/210122/#anticipated-errors) and LKE has `kube-proxy` and `calico`.

Is this not a result of an understandable but naive configurations by these providers? The assumption is that each node in clusters that they provision|manage is a vanilla kubelet.

Could we not submit feature requests against these providers soliciting them to revise the manifests of infrastructural sets so that they have more precise tolerations on the sets and taints on their nodes? Yes, it outsources effort to others who may be less inclined to help but it may also help them understand a shortcoming in their current approach.

**Comment by kflansburg** [comment]
@DazWilkin Can you obtain tolerations and node selectors for those DaemonSets?

My current theory is that we are incorrectly setting the node label `kubernetes.io/os: linux`. This is what Kind looks for for `kube-proxy`. It appears that this field is used to determine what flavor of container is supported (which doesnt seem totally correct).

**Comment by DazWilkin** [comment]
IIUC (!?) Krustlet is behaving correctly but, when it is added to one of these managed clusters, the cluster naively assumes another regular|vanilla kubelet has joined and it tries to deploy its infrastructural Pods to it e.g. `kube-proxy`.

As @thomastaylor312 points out above, this is because these Pods are configured by e.g. DaemonSets that have ~wildcard tolerations which result in them being deployed to all nodes regardless of whether the node (e.g. Krustlet) has precise taints.

This appears to me to be a problem with the managed services **not** Krustlet and I propose we raise this with them in order that they may have an epiphany on their overly broad settings and refine them (even if not only for Krustlet but by way of "specificity beating generality").

I'm playing with LKE now and:

```bash
kubectl get daemonset/kube-proxy \
--namespace=kube-system \
--output=jsonpath="{.spec.template.spec.tolerations}" | jq .
[
  {
    "key": "CriticalAddonsOnly",
    "operator": "Exists"
  },
  {
    "operator": "Exists"
  }
]
```
IIUC, that's effectively a wildcard toleration that means any node (and the node can't taint it off).

**Comment by kflansburg** [comment]
Yes, I understand the issue with the toleration. My point is that, at least in the case of Kind, there _is_ a node selector that we can use to avoid scheduling. I'm interested in seeing if the same selector is being used by these other vendors. There may be a convention here that we are not understanding / respecting (i.e. that architecture is indicated with this label rather than a taint).

**Comment by DazWilkin** [comment]
Apologies

**Comment by DazWilkin** [comment]
> @DazWilkin Can you obtain tolerations and node selectors for those DaemonSets?

Yes, I'm going to go walk with my dog and I'll do this Monday.

**Comment by DazWilkin** [comment]
For Digital Ocean:

```bash
for SET in "cilium" "csi-do-node" "do-node-agent" "kube-proxy"
do
  kubectl get daemonset/${SET} \
  --namespace=kube-system \
  --output=jsonpath="{.spec.template.spec.tolerations}" \
  | jq -r
done

cilium
[
  {
    "operator": "Exists"
  }
]
csi-do-node
[
  {
    "effect": "NoExecute",
    "operator": "Exists"
  },
  {
    "effect": "NoSchedule",
    "operator": "Exists"
  },
  {
    "key": "CriticalAddonsOnly",
    "operator": "Exists"
  }
]
do-node-agent
[
  {
    "operator": "Exists"
  }
]
kube-proxy
[
  {
    "effect": "NoExecute",
    "operator": "Exists"
  },
  {
    "effect": "NoSchedule",
    "operator": "Exists"
  },
  {
    "key": "CriticalAddonsOnly",
    "operator": "Exists"
  }
]
```
And:

The nodes have **no** taints by default. The following is empty:

```bash
kubectl get nodes --output=jsonpath="{.items[*].spec.taints}"
```

**Comment by DazWilkin** [comment]
Problems creating clusters on Linode today; will try again tomorrow

**Comment by DazWilkin** [comment]
For Linode Kubernetes Engine:

```bash
for SET in "calico-node" "csi-linode-node" "kube-proxy"
do
  echo ${SET}
  kubectl get daemonset/${SET} \
  --namespace=kube-system \
  --output=jsonpath="{.spec.template.spec.tolerations}" \
  | jq -r
done

calico-node
[
  {
    "effect": "NoSchedule",
    "operator": "Exists"
  },
  {
    "key": "CriticalAddonsOnly",
    "operator": "Exists"
  },
  {
    "effect": "NoExecute",
    "operator": "Exists"
  },
  {
    "effect": "NoSchedule",
    "key": "node.kubernetes.io/not-ready",
    "operator": "Exists"
  }
]
csi-linode-node
[]
kube-proxy
[
  {
    "key": "CriticalAddonsOnly",
    "operator": "Exists"
  },
  {
    "operator": "Exists"
  }
]
```
And:

The nodes have **no** taints by default. The following is empty:

```bash
kubectl get nodes --output=jsonpath="{.items[*].spec.taints}"
```

**Comment by kflansburg** [comment]
From @DazWilkin : 

DO DaemonSet `nodeSelector`:
```json
cilium
csi-do-node
do-node-agent
{
  "beta.kubernetes.io/os": "linux"
}
kube-proxy
```

LKE (Linode) DaemonSet `nodeSelector`:
```json
calico-node
{
  "beta.kubernetes.io/os": "linux"
}
csi-linode-node
kube-proxy
```

**Comment by kflansburg** [comment]
Our proposed fix was already applied in #468. Any Kubernetes flavors that still suffer from this problem are not using the node selector `beta.kubernetes.io/os=linux`, and we will have to determine another fix.
